### ISSUE TITLE
Port from container are not exposed by kubedock

### ISSUE THREAD

**Issue by wgebis** [joyrex2001/kubedock#3]
Hello there,

when I'm running e.g. mongodb as a testcontainers image, port 27017 is not exposed by the kudedock, `kubectl` says that:
```
    Container ID:  containerd://f08607808925b030a57c604f02904ce8f74c02fd9fdf43fb317281c21f6f06e0
    Image:         mongo:4.4.10
    Image ID:      docker.io/library/mongo@sha256:2821997cba3c26465b59cc2e863b940d21a58732434462100af10659fc0d164f
    Port:          27017/TCP
    Host Port:     0/TCP
    Args:
      --replSet
      docker-rs
```

Testcontainers test suite reports log:
```
org.springframework.dao.DataAccessResourceFailureException: Timed out after 30000 ms while waiting for a server that matches WritableServerSelector. Client view of cluster state is {type=UNKNOWN, servers=[{address=docker:27017, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketOpenException: Exception opening socket}, caused by {java.net.ConnectException: Connection refused}}]; nested exception is com.mongodb.MongoTimeoutException: Timed out after 30000 ms while waiting for a server that matches WritableServerSelector. Client view of cluster state is {type=UNKNOWN, servers=[{address=docker:27017, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketOpenException: Exception opening socket}, caused by {java.net.ConnectException: Connection refused}}]
```

`docker` is the container alias, port `2475` is exposed and docker API is available.  

Ports of target containers are not accessible from the container which uses kubedock. Do anyone know the reason of this?

Cheers,
W

**Comment by joyrex2001** [comment]
Did you run kubedock with port-forwarding `--port-forward` or `--reverse-proxy`? Port-forwarding is typical useful for running locally against a remote k8s cluster, while reverse-proxy is more stable/faster and can be used if the tests are run in the same cluster as kubedock.

**Comment by wgebis** [comment]
Thanks for response. Yeah, it helps, but know I'm trying to figured out randomly testcontainer failures. 
BTW. Is there any way to access private repos from `kubedock` container? Now it raises 403, looks like it doesn't honor `DOCKER_AUTH_CONFIG` env variable. 
Kubedocs seems to fill the gap between dind requirements and k8s env but so far it's so painful...

**Comment by wgebis** [comment]
@joyrex2001 Looks that I have stable env. Many thanks for help. 👍
One issue is still open, as mentioned in previous post, is there any way to inject/honor DOCKER_AUTH_CONFIG which provides credential for private registries with images created via kubedock?

**Comment by joyrex2001** [comment]
Awesome :-)

Since kubedock is orchestrating the containers in kubernetes, the docker_auth_config should be configured there as well. This is done with image pull secrets. See also: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/

**Comment by wgebis** [comment]
I'm aware of it, but pod has to be known which kubernetes secret objects should use in order to make auth (snippet below).
```
apiVersion: v1
kind: Pod
metadata:
  name: private-reg
spec:
  containers:
  - name: private-reg-container
    image: <your-private-image>
  imagePullSecrets:
  - name: regcred
```

For now, images created by kubedock does not use imagePullSecrets at all, but has to in order access images from private repo.

**Comment by wgebis** [comment]
@joyrex2001  One more think I found, when I have k8s cluster with two nodes, and scenario when kubedock is running on node A and testcontainers containers ware scheduled  on nodeB, sometimes there are a lot of communication issues between the containers - tests fails randomly. 
How it could be handled? Do you know the possible reasons?

**Comment by joyrex2001** [comment]
> For now, containers created by kubedock does not use imagePullSecrets at all, but has to in order access images from private repo.

Configuring specific pull secrets is not yet supported, this doesn't look like a huge implementation effort though.

> @joyrex2001 One more think I found, when I have k8s cluster with two nodes, and scenario when kubedock is running on node A and testcontainers containers ware scheduled on nodeB, sometimes there are a lot of communication issues between the containers - tests fail randomly.

My guess is that you use port-forwarding, which is not that stable as this tunnels over the k8s api. If both kubedock and the tests using testcontainers are running in the same cluster, reverse-proxy is recommended (like done in the tekton example).

**Comment by wgebis** [comment]
I'll stay tuned waiting for support private repos 👍 and I'm going to try to handle random fails. 
Thanks you, cheers!

**Comment by wgebis** [comment]
@joyrex2001 
I also observe from time to time testcontainers starts running the tests when mongo containers are not started already - it fails at the beginning of the test run. Is it possible that K8S API is called with no blocking manner like fire-and-forget? In opposite to Docker API it could be the difference.

**Comment by joyrex2001** [comment]
> I also observe from time to time testcontainers starts running the tests when mongo containers are not started already - it fails at the beginning of the test run. Is it possible that K8S API is called with no blocking manner like fire-and-forget? In opposite to Docker API it could be the difference.

This is similar to docker; the container is started, but it might not be ready yet. In the testcontainers framework this is usually managed by 'waitingFor' expressions, to make sure the test code is not executed before a certain (ready) state has been met.

**Comment by joyrex2001** [comment]
Configuring pull-secrets has been released in 0.8.0. The ```--image-pull-secrets``` argument has been added, which takes a comma-separated list of pull-secrets.

**Comment by wgebis** [comment]
Thanks! I’ll upgrade my env soon, cheers! 👍
